Update Sparkle to version 1.24.0 to support Apple Silicon

mokagio opened the pull request:
**Update: Superseded by #824** 

---

This builds on top of #802 and only upgrades Sparkle to allow the internal Simplenote target to build for Apple Silicon.

I moved it in an isolated PR because of the huge diff. You'll notice there is not change other than the Sparkle files, so if the Simplenote target builds, runs, and archives as normal for you, that should be enough.

mokagio committed:
Update Sparkle to version 1.24.0 to support Apple Silicon

AliSoftware commented:
I'll wait for https://github.com/Automattic/Automattic-Tracks-iOS/pull/166 then #802 before considering testing this one, but in the meantime, quick question: any reason why we include Sparkle manually and commit it into git (`External/*`), as opposed to include it via CocoaPods or something?

I don't know if Sparkle provides a podspec (even if it doesn't we might consider creating a local `.podspec` file if needed), but if they were to at least distribute it in a way that would avoid us having to commit it, it would (1) reduce the repo clone size (2) make diffs easier (3) reduce noice in PRs like this one. 🤷

mokagio commented:
> any reason why we include Sparkle manually and commit it into git (`External/*`), as opposed to include it via CocoaPods or something?

The setup pre-dates me. I'm not sure if there was a specific reason or if maybe CocoaPods wasn't supported at the time.

> I don't know if Sparkle provides a podspec (even if it doesn't we might consider creating a local .podspec file if needed), but if they were to at least distribute it in a way that would avoid us having to commit [...]

Totally. I didn't go down that route to keep this PR focused on making the build work vs. changing integration method. While that might not have interfered with anything, I feel like I already changed enough infrastructure already 😅

mokagio commented:
Moved to 2.7 as per [this comment in the base PR](https://github.com/Automattic/simplenote-macos/pull/802#issuecomment-765034695).

jkmassel committed:
Build for Sparkle in debug non-App Store builds

This will prevent it from randomly breaking when we try to do a production build

jkmassel committed:
Fix Sparkle

jkmassel reviewed (changes requested):
This was actually a breaking change, but it wasn't detected as such because we only built the Sparkle libraries for release builds previously, so CI didn't catch it. I've added two commits: 

- One to cause the build to fail due to the change
- One to fix Sparkle

Because it is a breaking change, IMHO this is a good opportunity to adopt Sparkle as a CocoaPod (since we'll have to test it extensively anyway).

Interested to hear your thoughts since you've already looked at it @AliSoftware.

mokagio committed:
Update Sparkle feed URL to use App Center format

This was needed anyways but became required as, in moving from version
1.22 manually added to the project to 1.24 linked via CocoaPods, the
original Hockey URL stopped working.

mokagio committed:
Link Sparkle 1.24 via CocoaPods instead of manually

We decided to bite the bullet and move away from manually downloading,
copying, and tracking the framework to simplify our setup.

See discussion in https://github.com/Automattic/simplenote-macos/pull/803.

I considered using Swift Package Manager, which is the preferred
dependency manager for Simplenote iOS and macOS but saw two open issues
in the Sparkle repo that mentioned the need for workarounds.

- https://github.com/sparkle-project/Sparkle/issues/1689
- https://github.com/sparkle-project/Sparkle/issues/1701

It's possible that the installation would have been straightforward for
us, but I decided to stick with the tried-and-tested, one-step CocoaPods
setup instead.

mokagio committed:
Merge branch 'release/2.6-apple-silicon' into release/2.6-apple-silicon-sparkle

Conflicts:
- `Simplenote/SimplenoteAppDelegate.m`

mokagio commented on Simplenote.xcodeproj/project.pbxproj:
```diff
@@ -1855,19 +1809,6 @@
 			shellPath = /bin/sh;
 			shellScript = "plist=\"${TARGET_BUILD_DIR}/${INFOPLIST_PATH}\"\nyear=$(date \"+%Y\")\ncopyright=\"Copyright © $year Automattic. All rights reserved.\"\n\n# Set the version numbers in the buildPlist\n#\n# Temporarily disabled! It's failing the build after updating to support Apple Silicon.\n#\n# Unfortunately, a number of changes were required to make the build pass before getting to this point ...
 		};
-		B51D44911945FEFD00496889 /* ShellScript */ = {
-			isa = PBXShellScriptBuildPhase;
-			buildActionMask = 2147483647;
-			files = (
-			);
-			inputPaths = (
-			);
-			outputPaths = (
-			);
-			runOnlyForDeploymentPostprocessing = 0;
-			shellPath = /bin/sh;
-			shellScript = "# Codesign\n#\nIDENTITY=\"Developer ID Application: Automattic, Inc.\"\nSPARKLE_BIN=\"${SRCROOT}/External/Sparkle/bin\"\nCODESIGN_TOOL=\"${SPARKLE_BIN}/codesign_embedded_executable\"\nCODESIGN_HARDENEDRUNTIME_CMD=\"codesign --force -o runtime -s\"\n\n# XPC\n#\nXPC_SOURCE=\"${SRCROOT}/External/Sparkle/XPCServices\"\nXPC_TARGET=\"${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/../ ...
-		};
```
The CocoaPods should make this unnecessary. See https://github.com/sparkle-project/Sparkle/issues/709.

jkmassel replied:
Yep, that's a huge win for sure!

mokagio commented on Simplenote/Simplenote-Info-Hockey.plist:
```diff
@@ -53,7 +53,7 @@
 	<string>SPApplication</string>
 	<key>SUFeedURL</key>
-	<string>https://rink.hockeyapp.net/api/2/apps/ede9c49346946c51a1db1db167b49892</string>
+	<string>https://api.appcenter.ms/v0.1/public/sparkle/apps/ede9c493-4694-6c51-a1db-1db167b49892</string>
```
See #639. Without this, the update check would fail with 

```
[Sparkle] Error: An error occurred in retrieving update information. Please try again later. An error occurred while parsing the update feed. (URL (null))
[Sparkle] Error: An error occurred while parsing the update feed. (null) (URL (null))
[Sparkle] Error: Line 6: Opening and ending tag mismatch: hr line 5 and body
 (null) (URL (null))
```

mokagio commented:
🤔 I'm getting the update dialog, the new version is downloaded correctly, but then I end up stuck on the "Install and Relaunch" dialog

<img width="512" alt="Screen Shot 2021-01-27 at 4 25 29 pm" src="https://user-images.githubusercontent.com/1218433/105947075-91baf900-60bc-11eb-849b-bd0ebb3899b0.png">

mokagio committed:
Merge branch 'release/2.6-apple-silicon' into release/2.6-apple-silicon-sparkle

jkmassel commented:
> I end up stuck on the "Install and Relaunch" dialog

I wonder if it has anything to do with code signing? Something should show up in `Console.app` if gatekeeper is blocking a process (I think...)

mokagio commented:
@jkmassel you might be onto something there.

Yesterday, I looked into `Console.app` but couldn't find anything. I found that suspicious, so today I did some reading to learn how to use the tool properly and was able to find the logs Simplenote macOS emits.

![image](https://user-images.githubusercontent.com/1218433/106087457-26366180-6178-11eb-83d0-b6af40f24473.png)


Among those, I found these:

```
Error registering stapled ticket: Error Domain=NSOSStatusErrorDomain Code=-67048 "(null)"
Warning: could not remove quarantine metadata from /Users/gio/Library/Containers/com.automattic.SimplenoteMac.beta/Data/Library/Caches/com.automattic.SimplenoteMac.beta/org.sparkle-project.Sparkle/Simplenote (Autoupdate).app: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" UserInfo={NSLocalizedDescription=Failed to remove file quarantine on Simplenote (Autoupdate).app.}. This may not be a problem, the update process will continue.
```

I might run out of time to follow this up today, but it's a promising lead to get to the bottom of the issue.

mokagio commented:
Closing in favor of #824.
